.evergreen/happy-eyeballs-server.py

```diff
@@ -0,0 +1,135 @@
+    print(f'{PREFIX}: listening for control connections on {args.control}', file=sys.stderr)
+    async with srv:
+        await shutdown.wait()
+    print(f'{PREFIX}: all done', file=sys.stderr)
+
+async def on_control_connected(reader: asyncio.StreamReader, writer: asyncio.StreamWriter, shutdown: asyncio.Event):
+    # Read the control request byte
+    data = await reader.readexactly(1)
+    if data == b'\x04':
+        print(f'{PREFIX}: ========================', file=sys.stderr)
+        print(f'{PREFIX}: request for delayed IPv4', file=sys.stderr)
+    elif data == b'\x06':
+        print(f'{PREFIX}: ========================', file=sys.stderr)
+        print(f'{PREFIX}: request for delayed IPv6', file=sys.stderr)
+    elif data == b'\xF0':
+        writer.write(b'\x01')
+        await writer.drain()
+        writer.close()
+        await writer.wait_closed()
+        return
+    elif data == b'\xFF':
+        print(f'{PREFIX}: shutting down', file=sys.stderr)
+        writer.close()
+        await writer.wait_closed()
+        shutdown.set()
+        return
+    else:
+        print(f'Unexpected control byte: {data}', file=sys.stderr)
+        exit(1)
+    
+    # Create the test servers but do not yet start accepting connections
+    connected = asyncio.Event()
+    on_ipv4_connected = lambda reader, writer: on_test_connected('IPv4', writer, b'\x04', connected)
+    on_ipv6_connected = lambda reader, writer: on_test_connected('IPv6', writer, b'\x06', connected)
+    srv4 = await asyncio.start_server(on_ipv4_connected, 'localhost', family=socket.AF_INET, start_serving=False)
+    srv6 = await asyncio.start_server(on_ipv6_connected, 'localhost', family=socket.AF_INET6, start_serving=False)
+    ipv4_port = srv4.sockets[0].getsockname()[1]
+    ipv6_port = srv6.sockets[0].getsockname()[1]
+    print(f'{PREFIX}: open for IPv4 on {ipv4_port}', file=sys.stderr)
+    print(f'{PREFIX}: open for IPv6 on {ipv6_port}', file=sys.stderr)
+
+    # Reply to control request with success byte and test server ports
+    writer.write(b'\x01')
+    writer.write(ipv4_port.to_bytes(2, 'big'))
+    writer.write(ipv6_port.to_bytes(2, 'big'))
+    await writer.drain()
+    writer.close()
+    await writer.wait_closed()
+
+    # Start test servers listening in parallel
+    await asyncio.wait([
+        asyncio.create_task(test_listen('IPv4', srv4, data == b'\x04', connected)),
+        asyncio.create_task(test_listen('IPv6', srv6, data == b'\x06', connected)),
+    ])
+
+    # Wait for the test servers to shut down
+    srv4.close()
+    srv6.close()
+    await asyncio.wait([
+        asyncio.create_task(srv4.wait_closed()),
+        asyncio.create_task(srv6.wait_closed()),
+    ])
+    
+    print(f'{PREFIX}: connection complete, test ports closed', file=sys.stderr)
+    print(f'{PREFIX}: ========================', file=sys.stderr)
+
+async def test_listen(name: str, srv, delay: bool, connected: asyncio.Event):
```

Comment:
Fun fact: `srv` can't have a type annotation because the version of python3 on some Evergreen hosts is old enough that the correct type (`asyncio.Server`) isn't public.